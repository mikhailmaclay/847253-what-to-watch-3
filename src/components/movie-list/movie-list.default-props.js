import {Config} from '../../consts';

export default {
  renderedMovieCards: Config.MOVIE_CARDS_COUNT_AT_START // withCounter
};
