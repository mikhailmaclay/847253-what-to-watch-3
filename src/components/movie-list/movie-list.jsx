// Libraries
import React from 'react';
// PropTypes
import propTypes from './movie-list.prop-types';
import defaultProps from './movie-list.default-props';
// Constants and utils
import {Config} from '../../consts';
// Components
import MovieCard from '../movie-card/movie-card';
import Button from '../button/button';
// HOCs
import withCounter from '../../hocs/with-counter';

function MovieList({movies, count: itemsCount, incrementCounter}) {
  const renderMovieCards = () => (
    movies.slice(0, itemsCount).map((movie) => <MovieCard key={movie.id} {...movie}/>)
  );

  const handleButtonClick = () => {
    incrementCounter(Config.MOVIE_CARDS_COUNT_BY_LOAD);
  };

  return (
    <>
      <div className="catalog__movie-list">
        {renderMovieCards()}
      </div>
      {itemsCount < movies.length &&
        <div className="catalog__more">
          <Button className="catalog__button" onClick={handleButtonClick}>Show more</Button>
        </div>
      }
    </>
  );
}

MovieList.propTypes = propTypes;
MovieList.defaultProps = defaultProps;

const MovieListWithCounter = withCounter(MovieList, Config.MOVIE_CARDS_COUNT_AT_START);

export default MovieListWithCounter;
export {MovieList};
