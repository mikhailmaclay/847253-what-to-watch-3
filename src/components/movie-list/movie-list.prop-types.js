import PropTypes from 'prop-types';

export default {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired
  })).isRequired,
  count: PropTypes.number.isRequired,
  incrementCounter: PropTypes.func.isRequired,
  onMouseOver: PropTypes.func
};
