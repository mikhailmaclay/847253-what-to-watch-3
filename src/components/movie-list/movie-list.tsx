// Libraries
import * as React from 'react';
// Constants and utils
import {Config} from '../../constants/consts';
import createHOC from '../../utils/components/create-hoc';
// Components
import MovieCardWrapped from '../movie-card/movie-card';
import Button from '../button/button';
//
import {Movie} from '../../types';

interface Props {
  movies: Movie[];
  withCounter?: {
    renderedMovieCards: {
      value: number;
      increment: (value: number) => void;
    };
  };
}

function MovieList(props: Props) {
  const {movies, withCounter} = props;

  const renderMovieCards = (count) => (
    movies.slice(0, count).map((movie) => <MovieCardWrapped key={movie.id} {...movie}/>)
  );

  if (!withCounter) { // if not a withCounter HOC
    return (
      <div className="catalog__movie-list">
        {renderMovieCards(Config.MOVIE_CARDS_COUNT_AT_START)}
      </div>
    );
  }

  const handleButtonClick = () => {
    withCounter.renderedMovieCards.increment(Config.MOVIE_CARDS_COUNT_BY_LOAD);
  };

  return (
    <>
      <div className="catalog__movie-list">
        {renderMovieCards(withCounter.renderedMovieCards.value)}
      </div>
      {withCounter.renderedMovieCards.value < movies.length &&
        <div className="catalog__more">
          <Button className="catalog__button" onClick={handleButtonClick}>Show more</Button>
        </div>
      }
    </>
  );
}

const MovieListMemo = React.memo(MovieList);
const MovieListWrapped = createHOC(`withCounter`, {counterName: `renderedMovieCards`, initialValue: Config.MOVIE_CARDS_COUNT_AT_START, isResettableOnUpdate: true})(MovieListMemo);

export default MovieListWrapped;
export {MovieList, MovieListMemo};
