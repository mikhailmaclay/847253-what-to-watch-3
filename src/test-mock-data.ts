import {User, Movie, Review, TeamMember, Notification} from './types';

export const mockUser: User = {
  id: 1,
  email: `JohnnyMnemonyc@mail.ru`,
  name: `Johnny`,
  avatar: `https://htmlacademy-react-3.appspot.com/wtw/static/avatar/4.jpg`,
  myList: null
};

export const mockReview: Review = {
  id: 1,
  user: {id: mockUser.id, fullName: mockUser.name},
  date: `2020-03-24T08:26:50.139Z`,
  rating: 10,
  text: `It was really awesome!`
};

export const mockDirector: TeamMember = {
  fullName: `Wes Anderson`,
  role: `Director`
};

export const mockActor: TeamMember = {
  fullName: `Adrien Brody`,
  role: `Actor`
};

export const mockMovie: Movie = {
  id: 1,
  name: `Dardjeeling Limited`,
  genre: `Adventure`,
  releaseDate: `2007`,
  description: [
    `A year after their father\`s funeral, three brothers travel across India by train in an attempt to bond with each other.`
  ],
  runTime: 91,
  team: [
    mockDirector,
    mockActor
  ],
  poster: `https://htmlacademy-react-3.appspot.com/wtw/static/film/poster/Dardjeeling_Limited.jpg`,
  images: [
    `https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/dardjeeling_limited.jpg`
  ],
  backgrounds: [
    `#AD9F8B`,
    `https://htmlacademy-react-3.appspot.com/wtw/static/film/background/Dardjeeling_Limited.jpg`
  ],
  rating: 3.6,
  scoresCount: 165106,
  preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  video: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  isInMyList: false,
  reviews: null
};

export const mockNotification: Notification = {
  title: `Test`,
  content: `Testing message.`
};
