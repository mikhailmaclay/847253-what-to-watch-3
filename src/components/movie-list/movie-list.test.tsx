// Libraries
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// Components
import {MovieList} from './movie-list';
//
import {mockMovie} from '../../test-mock-data';

describe(`<MovieList/>`, () => {
  const props = {
    movies: [mockMovie]
  };

  it(`Should render correctly`, () => {
    const result = renderer.create(
        <MovieList {...props}/>
    ).toJSON();

    expect(result).toMatchSnapshot();
  });
});
