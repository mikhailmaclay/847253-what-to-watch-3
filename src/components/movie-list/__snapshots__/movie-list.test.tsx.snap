// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieList/> Should render correctly 1`] = `
<div
  className="catalog__movie-list"
>
  <article
    className="small-movie-card catalog__movie-card"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <a
      className="small-movie-card__image"
      href="/films/1"
      onClick={[Function]}
    >
      <img
        alt="Dardjeeling Limited"
        height="175"
        src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/dardjeeling_limited.jpg"
        width="280"
      />
    </a>
    <h3
      className="small-movie-card__title"
    >
      <span
        className="small-movie-card__link"
      >
        Dardjeeling Limited
      </span>
    </h3>
  </article>
</div>
`;
